test/test-tcp-exportimport.c

```diff
@@ -0,0 +1,237 @@
+
+    r = uv_tcp_init(uv_default_loop(), &conn->conn);
+    ASSERT(r == 0);
+
+    r = uv_ip4_addr("127.0.0.1", TEST_PORT, &addr);
+    ASSERT(r == 0);
+
+    r = uv_tcp_connect(&conn->conn_req,
+                       (uv_tcp_t*)&conn->conn,
+                       (struct sockaddr*) &addr,
+                       connect_cb);
+    ASSERT(r == 0);
+
+    conn->conn.data = conn;
+  }
+}
+
+
+void on_parent_msg(uv_async_t* handle) {
+  int r;
+
+  ASSERT(dup_stream.type == UV_TCP);
+
+  /* Import the shared TCP server, and start listening on it. */
+  r = uv_tcp_init(parent.loop, &parent.server);
+  ASSERT(r == 0);
+
+  parent.server.data = &parent;
+
+  r = uv_import((uv_stream_t*)&parent.server,
+               (uv_stream_info_t*)&dup_stream);
+  ASSERT(r == 0);
+
+  r = uv_listen((uv_stream_t*)&parent.server, 12, on_connection);
+  ASSERT(r == 0);
+
+  /* Create a bunch of connections to get both servers to accept. */
+  make_many_connections();
+}
+
+
+void on_child_msg(uv_async_t* handle) {
+  ASSERT(!"no");
+}
+
+
+static void child_thread_entry(void* arg) {
+  int r;
+  int listen_after_write = *(int*) arg;
+  struct sockaddr_in addr;
+
+  r = uv_tcp_init(child.loop, &child.server);
+  ASSERT(r == 0);
+
+  child.server.data = &child;
+  
+  r = uv_ip4_addr("0.0.0.0", TEST_PORT, &addr);
+  ASSERT(r == 0);
+
+  r = uv_tcp_bind(&child.server, (struct sockaddr*) &addr, 0);
+  ASSERT(r == 0);
+
+  if (!listen_after_write) {
+    r = uv_listen((uv_stream_t*)&child.server, 12, on_connection);
+    ASSERT(r == 0);
+  }
+
+  r = uv_export((uv_stream_t*)&child.server,
+    (uv_stream_info_t*)&dup_stream);
+  ASSERT(r == 0);
+
+  r = uv_async_send(child.send_channel);
+  ASSERT(r == 0);
+
+  if (listen_after_write) {
+    r = uv_listen((uv_stream_t*)&child.server, 12, on_connection);
+    ASSERT(r == 0);
+  }
+
+  r = uv_run(child.loop, UV_RUN_DEFAULT);
+  ASSERT(r == 0);
+
+  ASSERT(child.connection_accepted == 1);
+  ASSERT(child.close_cb_called == 3);
+}
+
+
+static void run_tcp_exportimport_test(int listen_after_write) {
+  int r;
+
+  parent.send_channel = &send_channel;
+  parent.recv_channel = &recv_channel;
+  child.send_channel = &recv_channel;
+  child.recv_channel = &send_channel;
+
+  parent.loop = uv_default_loop();
+  child.loop = uv_loop_new();
+  ASSERT(child.loop);
+
+  r = uv_async_init(parent.loop, parent.recv_channel, on_parent_msg);
+  ASSERT(r == 0);
+  parent.recv_channel->data = &parent;
+
+  r = uv_async_init(child.loop, child.recv_channel, on_child_msg);
+  ASSERT(r == 0);
+  child.recv_channel->data = &child;
+
+  r = uv_thread_create(&child.thread, child_thread_entry, &listen_after_write);
+  ASSERT(r == 0);
+
+  r = uv_run(parent.loop, UV_RUN_DEFAULT);
+  ASSERT(r == 0);
+
+  ASSERT(parent.connection_accepted == 1);
+  ASSERT(parent.close_cb_called == 3);
+
+  r = uv_thread_join(&child.thread);
+  ASSERT(r == 0);
+
+  /* Satisfy valgrind. Maybe we should delete the default loop from the
+   * test runner.
+   */
```

Comment:
```suggestion
  /* Satisfy valgrind. */
```
We settled on the `MAKE_VALGRIND_HAPPY()` macro for the default event loop a long time ago. :-)